File: assets/js/product/product.js
'use strict';

module.exports = function(data){
  data = data || {};
  this.id = m.prop(data.id || 0);
  this.name = m.prop(data.name || '');
  this.slug = m.prop(data.slug || '');
  this.content = m.prop(data.content || '');
  this.image = m.prop(data.image || '');
  this.published = m.prop(data.published || true);
  this.price = m.prop(data.price || null);
  this.category_name = m.prop((data.category) ? data.category.name : '');
  this.category_id = m.prop(data.category_id || 0);
  this.meta_description = m.prop(data.meta_description || '');
  this.meta_keywords = m.prop(data.meta_keywords || '');
}

File: models/product.go
package models

import (
	"fmt"
	"html/template"
)

//Product type contains product info
type Product struct {
	ID              uint    `json:"id"`
	Name            string  `json:"name"`
	Slug            string  `json:"slug"`
	Content         string  `json:"content"`
	Image           string  `json:"image"`
	Published       bool    `json:"published"`
	Price           float32 `json:"price,string"`
	MetaDescription string  `json:"meta_description"`
	MetaKeywords    string  `json:"meta_keywords"`
	//relations
	CategoryID uint     `gorm:"index" json:"category_id"`
	Category   Category `json:"category"`
	Images     []Image  `json:"images"`
}

func (product *Product) URL() string {
	return fmt.Sprintf("/product/%d-%s", product.ID, product.Slug)
}

func (product *Product) EditURL() string {
	return fmt.Sprintf("/admin#/products/%d", product.ID)
}

func (product *Product) HTMLContent() template.HTML {
	return template.HTML(product.Content)
}

func (product *Product) BeforeCreate() (err error) {
	product.Slug = createSlug(product.Name)
	return
}

func (product *Product) BeforeSave() (err error) {
	product.Slug = createSlug(product.Name)
	return
}
